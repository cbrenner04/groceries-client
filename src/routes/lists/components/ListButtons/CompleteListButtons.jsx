import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { Refresh, Trash } from '../../../../components/ActionButtons';

function CompletedListButtons(props) {
  const userIsOwner = props.userId === props.list.owner_id;
  return (
    <ButtonGroup className="float-right">
      <Refresh
        handleClick={() => props.onListRefresh(props.list)}
        disabled={!userIsOwner}
        style={{ opacity: userIsOwner ? 1 : 0.3 }}
        data-test-id="complete-list-refresh"
      />
      <Trash
        handleClick={() => props.onListDeletion(props.list)}
        disabled={!userIsOwner}
        style={{ opacity: userIsOwner ? 1 : 0.3 }}
        data-test-id="complete-list-trash"
      />
    </ButtonGroup>
  );
}

CompletedListButtons.propTypes = {
  userId: PropTypes.number.isRequired,
  list: PropTypes.shape({
    owner_id: PropTypes.number.isRequired,
  }).isRequired,
  onListRefresh: PropTypes.func.isRequired,
  onListDeletion: PropTypes.func.isRequired,
};

export default CompletedListButtons;
