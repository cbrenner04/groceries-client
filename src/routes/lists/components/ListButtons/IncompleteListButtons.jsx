import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { Complete, Edit, Merge, Share, Trash } from '../../../../components/ActionButtons';

function IncompleteListButtons(props) {
  const userIsOwner = props.userId === props.list.owner_id;
  const userHasWritePermission = props.currentUserPermissions === 'write';

  return (
    <ButtonGroup className="float-right">
      <Complete
        handleClick={() => props.onListCompletion(props.list)}
        disabled={!userIsOwner}
        style={{ opacity: userIsOwner ? 1 : 0.3 }}
        data-test-id="incomplete-list-complete"
      />
      <Share
        to={`lists/${props.list.id}/users_lists`}
        disabled={!userHasWritePermission}
        style={{
          pointerEvents: userHasWritePermission ? 'auto' : 'none',
          opacity: userHasWritePermission ? 1 : 0.3,
        }}
        data-test-id="incomplete-list-share"
      />
      {props.multiSelect && <Merge handleClick={props.handleMerge} data-test-id="incomplete-list-merge" />}
      {!props.multiSelect && (
        <Edit
          to={`/lists/${props.list.id}/edit`}
          disabled={!userIsOwner}
          style={{
            pointerEvents: userIsOwner ? 'auto' : 'none',
            opacity: userIsOwner ? 1 : 0.3,
          }}
          data-test-id="incomplete-list-edit"
        />
      )}
      <Trash handleClick={() => props.onListDeletion(props.list)} data-test-id="incomplete-list-trash" />
    </ButtonGroup>
  );
}

IncompleteListButtons.propTypes = {
  userId: PropTypes.number.isRequired,
  list: PropTypes.shape({
    id: PropTypes.number.isRequired,
    owner_id: PropTypes.number.isRequired,
  }).isRequired,
  onListCompletion: PropTypes.func.isRequired,
  onListDeletion: PropTypes.func.isRequired,
  currentUserPermissions: PropTypes.string.isRequired,
  multiSelect: PropTypes.bool.isRequired,
  handleMerge: PropTypes.func.isRequired,
};

export default IncompleteListButtons;
