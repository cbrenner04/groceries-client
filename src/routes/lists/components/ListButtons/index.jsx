import React from 'react';
import PropTypes from 'prop-types';

import CompletedListButtons from './CompleteListButtons';
import IncompleteListButtons from './IncompleteListButtons';
import PendingListButtons from './PendingListButtons';

function ListButtons(props) {
  const acceptedListButtons = props.list.completed ? (
    <CompletedListButtons
      userId={props.userId}
      list={props.list}
      onListRefresh={props.onListRefresh}
      onListDeletion={props.onListDeletion}
    />
  ) : (
    <IncompleteListButtons
      userId={props.userId}
      list={props.list}
      onListCompletion={props.onListCompletion}
      onListDeletion={props.onListDeletion}
      currentUserPermissions={props.currentUserPermissions}
    />
  );

  return props.accepted ? (
    acceptedListButtons
  ) : (
    <PendingListButtons
      list={props.list}
      onListAcceptance={props.onListAcceptance}
      onListRejection={props.onListRejection}
    />
  );
}

ListButtons.propTypes = {
  userId: PropTypes.number.isRequired,
  list: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    created_at: PropTypes.string.isRequired,
    completed: PropTypes.bool.isRequired,
    users_list_id: PropTypes.number,
    owner_id: PropTypes.number.isRequired,
    refreshed: PropTypes.bool,
  }).isRequired,
  accepted: PropTypes.bool.isRequired,
  onListDeletion: PropTypes.func.isRequired,
  onListCompletion: PropTypes.func.isRequired,
  onListRefresh: PropTypes.func.isRequired,
  onListAcceptance: PropTypes.func.isRequired,
  onListRejection: PropTypes.func.isRequired,
  currentUserPermissions: PropTypes.string.isRequired,
};

export default ListButtons;
