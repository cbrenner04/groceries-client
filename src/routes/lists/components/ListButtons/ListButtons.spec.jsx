import React from 'react';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import ListButtons from './index';

describe('ListButtons', () => {
  let props;
  const renderListButtons = (props) => {
    const history = createMemoryHistory();
    return render(
      <Router history={history}>
        <ListButtons {...props} />
      </Router>,
    );
  };

  beforeEach(() => {
    props = {
      userId: 1,
      list: {
        id: 1,
        name: 'foo',
        type: 'GroceryList',
        created_at: new Date('05/27/2020').toISOString(),
        completed: false,
        users_list_id: 1,
        owner_id: 1,
        refreshed: false,
      },
      accepted: false,
      onListDeletion: jest.fn(),
      onListCompletion: jest.fn(),
      onListRefresh: jest.fn(),
      onListAcceptance: jest.fn(),
      onListRejection: jest.fn(),
      onListRemoval: jest.fn(),
      currentUserPermissions: 'write',
      multiSelect: false,
      selectedLists: [],
      handleMerge: jest.fn(),
    };
  });

  it('renders CompleteListButtons when list is accepted and completed', () => {
    props.accepted = true;
    props.list.completed = true;
    const { container, getByTestId, queryByTestId } = renderListButtons(props);

    expect(container).toMatchSnapshot();
    expect(getByTestId('complete-list-refresh')).toBeVisible();
    expect(getByTestId('complete-list-trash')).toBeVisible();
    expect(queryByTestId('incomplete-list-complete')).toBeNull();
    expect(queryByTestId('incomplete-list-share')).toBeNull();
    expect(queryByTestId('incomplete-list-edit')).toBeNull();
    expect(queryByTestId('incomplete-list-trash')).toBeNull();
    expect(queryByTestId('pending-list-accept')).toBeNull();
    expect(queryByTestId('pending-list-trash')).toBeNull();
  });

  it('renders IncompleteListButtons when list is accepted and not completed', () => {
    props.accepted = true;
    props.list.completed = false;
    const { container, getByTestId, queryByTestId } = renderListButtons(props);

    expect(container).toMatchSnapshot();
    expect(queryByTestId('complete-list-refresh')).toBeNull();
    expect(queryByTestId('complete-list-trash')).toBeNull();
    expect(getByTestId('incomplete-list-complete')).toBeVisible();
    expect(getByTestId('incomplete-list-share')).toBeVisible();
    expect(getByTestId('incomplete-list-edit')).toBeVisible();
    expect(getByTestId('incomplete-list-trash')).toBeVisible();
    expect(queryByTestId('pending-list-accept')).toBeNull();
    expect(queryByTestId('pending-list-trash')).toBeNull();
  });

  it('renders PendingListButtons when list is not accepted', () => {
    props.accepted = false;
    const { container, getByTestId, queryByTestId } = renderListButtons(props);

    expect(container).toMatchSnapshot();
    expect(queryByTestId('complete-list-refresh')).toBeNull();
    expect(queryByTestId('complete-list-trash')).toBeNull();
    expect(queryByTestId('incomplete-list-complete')).toBeNull();
    expect(queryByTestId('incomplete-list-share')).toBeNull();
    expect(queryByTestId('incomplete-list-edit')).toBeNull();
    expect(queryByTestId('incomplete-list-trash')).toBeNull();
    expect(getByTestId('pending-list-accept')).toBeVisible();
    expect(getByTestId('pending-list-trash')).toBeVisible();
  });
});
