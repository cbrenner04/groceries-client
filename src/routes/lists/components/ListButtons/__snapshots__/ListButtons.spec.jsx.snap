// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListButtons renders CompleteListButtons when list is accepted and completed 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="complete-list-refresh"
      style="opacity: 1;"
      type="button"
    >
      <i
        class="fa fa-redo fa-2x text-primary"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="complete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`ListButtons renders IncompleteListButtons when list is accepted and not completed 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="incomplete-list-complete"
      style="opacity: 1;"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="incomplete-list-share"
      style="pointer-events: auto; opacity: 1;"
      type="button"
    >
      <i
        class="fa fa-users fa-2x text-primary"
      />
    </button>
    <a
      class="p-0 mr-3"
      data-test-id="incomplete-list-edit"
      href="/lists/1/edit"
      style="pointer-events: auto; opacity: 1;"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </a>
    <button
      class="p-0 btn btn-link"
      data-test-id="incomplete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`ListButtons renders PendingListButtons when list is not accepted 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="pending-list-accept"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="pending-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;
